spinta/dimensions/param/ufuncs.py

```diff
@@ -1,10 +1,140 @@
+
+
+@ufunc.resolver(ParamBuilder, _Loop, Bind, name="getattr")
+def getattr_(env: ParamBuilder, _loop: _Loop, bind: Bind):
+    data = env.call('getattr', _loop.arg, bind)
+    if isinstance(data, Iterator):
+        for item in data:
+            if item:
+                yield item
+
+                if item != env.params[env.target_param]:
+                    env.params[env.target_param] = item
+                    args, kwargs = _loop.expr.resolve(env)
+                    for arg in args:
+                        yield from env.call('getattr', _Loop(_loop.expr, arg), bind)
+
+
+@ufunc.resolver(ParamBuilder, dict, Bind, name="getattr")
+def getattr_(env: ParamBuilder, data: dict, bind: Bind):
+    if bind.name in data:
+        yield data[bind.name]
+
+
+@ufunc.resolver(ParamBuilder, Expr)
+def loop(env: ParamBuilder, expr: Expr):
+    args, kwargs = expr.resolve(env)
+    for arg in args:
+        yield _Loop(expr, arg)
+
+
+@ufunc.resolver(ParamBuilder, Generator, Bind)
+def loop(env: ParamBuilder, generator: Generator, bind: Bind):
+    for item in generator:
+        data = env.call("getattr", item, bind)
+        yield data
+
+
+@ufunc.executor(ParamBuilder, NotAvailable)
+def select(env: ParamBuilder, na: NotAvailable) -> Any:
+    return env.this
+
+
+@ufunc.resolver(ParamLoader, list)
+def resolve_param(env: ParamLoader, params: list):
+    for parameter in params:
+        env.call("resolve_param", parameter)
+
+
+@ufunc.resolver(ParamLoader, Param)
+def resolve_param(env: ParamLoader, parameter: Param):
+    for i, (source, formula) in enumerate(zip(parameter.sources.copy(), parameter.formulas)):
+        if isinstance(source, str) and formula:
+            requires_model = env.call("contains_select", formula)
+            if requires_model:
+                new_name = source
+                if env.dataset and '/' not in source:
+                    new_name = '/'.join([
+                        env.dataset.name,
+                        source,
+                    ])
+                if commands.has_model(env.context, env.manifest, new_name):
+                    model = commands.get_model(env.context, env.manifest, new_name)
+                    parameter.sources[i] = model
+                elif source != new_name and commands.has_model(env.context, env.manifest, source):
+                    model = commands.get_model(env.context, env.manifest, source)
+                    parameter.sources[i] = model
+
+
+@ufunc.resolver(ParamLoader, object)
+def contains_select(env: ParamLoader, obj: object):
+    return False
+
+
+@ufunc.resolver(ParamLoader, Expr)
+def contains_select(env: ParamLoader, expr: Expr):
+    for arg in expr.args:
+        result = env.resolve(arg)
+        if isinstance(result, bool) and result is True:
+            return True
+    return False
+
+
+@ufunc.resolver(ParamLoader, object)
+def select(env: ParamLoader, any_: object):
+    return True
+
+
+@ufunc.resolver(ParamLoader, Expr)
```

Comment:
`Expr` is used, when you need to handle functions with dynamic arguments, for example `func(*args, **kwargs)`. But in this case we only have a single argument.

Instead `@ufunc.resolver(ParamLoader, Expr)` ir should be `@ufunc.resolver(paramLoader, types.GeneratorType)`.

And `loop` function implementation should look like this:

```python
@ufunc.resolver(ParamLoader, types.GeneratorType)
def loop(env: ParamLoader, rows: Iterable) -> Iterable:
    empty = False
    while not empty:
        empty = True
        for row in rows:
            empty = False
            yield row
```

Reply:
This is part of `ParamLoader`, it's main goal is validate and parse params at `link` time, to see if empty `read()` exists in the expression and if it does assume that `source` is a model and try to resolve it. `ParamBuilder` is used for looping through data.